lamp.py

```diff
@@ -0,0 +1,209 @@
+    env['NINJA_STATUS'] = '[%f/%t (%p) %es]'  # make the ninja output even nicer
+
+    logging.debug('Using environment: %s', env)
+    return env
+
+
+def cmake(toolchain_dir, cmdline_cmake_args, triplet_os, env):
+    cmake_cmd = ['cmake', '-B', 'build', '-G', 'Ninja']
+    # We set both the prefix path and the toolchain file here as a hack to allow cmake
+    # to find both shared and static libraries. vcpkg doesn't natively support a project
+    # using both.
+    cmake_prefix_path = os.path.join(toolchain_dir, 'installed/x64-{}-shared'.format(triplet_os))
+    cmake_toolchain_file = os.path.join(toolchain_dir, 'scripts/buildsystems/vcpkg.cmake')
+
+    cmake_cmd += [
+        '-DCMAKE_PREFIX_PATH={}'.format(cmake_prefix_path),
+        '-DCMAKE_TOOLCHAIN_FILE={}'.format(cmake_toolchain_file),
+        '-DVCPKG_TARGET_TRIPLET=x64-{}-static'.format(triplet_os),
+    ]
+
+    cmake_cmd += cmdline_cmake_args
+
+    logging.info('Running cmake: %s', cmake_cmd)
+    subprocess.run(cmake_cmd, env=env)
+
+
+def compile(env):
+    compile_cmd = ['ninja', '-C', 'build']
+    logging.info('Running ninja: %s', compile_cmd)
+    subprocess.run(compile_cmd, env=env)
+
+
+def install(env):
+    install_cmd = ['ninja', '-C', 'build', 'install']
+    logging.info('Running ninja install: %s', install_cmd)
+    subprocess.run(install_cmd, env=env)
+
+
+def cmake_test(env):
+    workdir = os.path.join(os.getcwd(), 'build')
+    ctest_cmd = [
+        'ctest',
+        '--label-exclude',
+        '(standalone|sharded|single_node_replset|three_node_replset)'
+    ]
+    subprocess.run(ctest_cmd, cwd=workdir, env=env)
+
+
+def parse_args(args, os_family):
+    parser = argparse.ArgumentParser(
+        description='Script for building genny',
+        epilog='Unknown positional arguments will be forwarded verbatim to the cmake invocation'
+    )
+
+    # Python can't natively check the distros of our supported platforms.
+    # See https://bugs.python.org/issue18872 for more info.
+    parser.add_argument('-d', '--linux-distro',
+                        choices=['ubuntu1804', 'arch', 'rhel7', 'amazon2', 'not-linux'],
+                        help='specify the linux distro you\'re on; if your system isn\'t available,'
+                             ' please contact us at #workload-generation')
+    parser.add_argument('-v', '--verbose', action='store_true')
+
+    subparsers = parser.add_subparsers(
+        dest='subcommand',
+        description='subcommands perform specific actions; make sure you run this script without any '
+                    'subcommand first to initialize the environment')
+    subparsers.add_parser(
+        'cmake-test', help='run cmake unit tests that don\'t connect to a MongoDB cluster')
+    subparsers.add_parser('compile', help='just run the compile step for genny')
+    subparsers.add_parser('install', help='just run the install step for genny')
+
+    known_args, unknown_args = parser.parse_known_args(args)
```

Comment:
It is worth adding a `--cmake-arg` as an append argument instead. (It would be used like `--cmake-arg "-DFOO=BAR"`.) It will make it much much clearer what is consumed and what is passed along.

Reply:
Hmmmm, that is true if you don't quote the cmake args, but we really should anyway.

Just making sure I wasn't crazy:
```
In [25]: parser = argparse.ArgumentParser()                                                                                                                  

In [26]: parser.add_argument("--cmake-arg", nargs=1, action='append')                                                                                        
Out[26]: _AppendAction(option_strings=['--cmake-arg'], dest='cmake_arg', nargs=1, const=None, default=None, type=None, choices=None, help=None, metavar=None)

In [27]: parser.parse_known_args('--cmake-arg "-DFOO=BAR" --cmake-arg "-DBAR"'.split())                                                                      
Out[27]: (Namespace(cmake_arg=[['"-DFOO=BAR"'], ['"-DBAR"']]), [])
```